modules/puppet_proxy/puppet_cache.rb

```diff
@@ -58,7 +58,16 @@ def read_from_cache environment
         cachefile = File.expand_path("cache_#{environment}.json", Proxy::Puppet::Plugin.settings.cache_location)
 
         if File.exist?(cachefile)
-          JSON.parse(File.read(cachefile))
+          as_hash = JSON.load(File.new(cachefile))
+          # ran into issues with deserialisation in older versions of json gem (ruby 1.8.7 + json 1.5.5),
+          # and opted for the deserialisation approach below.
+          # TODO: switch back to json library based deserialisation by implementing PuppetClass.json_create
+          # when support for Ruby 1.8.7 has been dropped.
+          as_hash.values.each do |puppet_module|
+            puppet_module['manifest'] =
+                puppet_module['manifest'].map { |klass| [Proxy::Puppet::PuppetClass.from_hash(klass.first)] }
```

Comment:
I get an `undefined method 'map' for nil:NilClass` error on this line.  The iteration doesn't look right to me, at least based on what appears in my JSON cache:

<pre>
[root@foreman6 ~]# json_reformat < /var/cache/foreman-proxy/cache_production.json                                                                                                                                                                                               
{
  "/etc/puppet/environments/production/modules": {
    "ntp": {
      "timestamp": 1436879142,
      "manifest": [
        [
          {
            "params": {
              "keys_file": "${$ntp::params::keys_file}",
</pre>

`as_hash` would be the top level hash with directories as keys, so `puppet_module` would be a hash of modules rather than a single one.

I think there's a mistake in the idempotency test I wrote which isn't taking into account the `cache[directory] = lcache` line (49) above, causing this to crash in real life.